conn.go

```diff
@@ -695,35 +714,31 @@ func (c *Conn) Handshake() Alert {
 	var alert Alert
 	if c.hState == nil {
 		logf(logTypeHandshake, "%s First time through handshake (or after stateless retry), setting up", label)
-		alert = c.HandshakeSetup()
-		if alert != AlertNoAlert {
+		alert := c.HandshakeSetup()
+		if alert != AlertNoAlert || (c.isClient && c.config.NonBlocking) {
 			return alert
 		}
-	} else {
-		logf(logTypeHandshake, "Re-entering handshake, state=%v", c.hState)
 	}
 
+	logf(logTypeHandshake, "Re-entering handshake, state=%v", c.hState)
```

Comment:
This log line seems wrong now that you've pulled it out of the `else` clause.  Need to change it to "(Re-)entering" or something